File: game/scripts/vscripts/game_events/akatsuki.lua
require("game_events/bidju_shinobi")
require("game_events/shinobi")



local function CreateDefender(bidju)

    local ent = Entities:FindByName(nil, "akatsuki_preview_npc_dota_custom_bidju_shukaku")
    local unit = CreateUnitByName("npc_dota_custom_cave", ent:GetAbsOrigin(), true, nil, nil, DOTA_TEAM_BADGUYS)
    unit.bidju_name = bidju:GetUnitName()
end

local function CaptureBidju(bidju)
    local entity = Entities:FindByName(nil, "akatsuki_preview_" .. bidju:GetUnitName())


    if entity.bidju_indicator ~= nil then
        ParticleManager:DestroyParticle(entity.bidju_indicator, false)
    end

    entity.bidju_indicator = ParticleManager:CreateParticle(
            "particles/env/bidju_capture_indicator.vpcf",
            PATTACH_ABSORIGIN,
            entity)

    ParticleManager:SetParticleControlEnt(entity.bidju_indicator, 0, entity, PATTACH_ABSORIGIN, "attach_hitloc", entity:GetAbsOrigin(), true) -- target


    CreateDefender(bidju)
end



function UncaptureBidju(name)

    local entity = Entities:FindByName(nil, "akatsuki_preview_" .. name)

    if entity.bidju_indicator ~= nil then
        ParticleManager:DestroyParticle(entity.bidju_indicator, false)
    end

    local unit = CreateUnitByName(name, Entities:FindByName(nil, "spawn_" .. name):GetAbsOrigin(), true, nil, nil, DOTA_TEAM_NEUTRALS)

    unit:StartGesture(ACT_DOTA_DISABLED)

    local duration = 5

    unit:AddNewModifier(owner, nil, "modifier_invulnerable", {
        duration = duration
    })

    Timers:CreateTimer(duration, function()
        unit:StartGesture(ACT_DOTA_IDLE)
    end)

end

function SetBidjuCaptured(bidju)
    CaptureBidju(bidju)
    if bidju.owner ~= nil then
        OnBidjuUncaptured(bidju.owner)
        TurnOffSummon(bidju.owner)
        bidju.owner.bidju = nil
        bidju.owner = nil
    end

end

File: game/scripts/vscripts/game_events/Akatsuki.ts
import {BidjuExtension} from "./bidju";


/** @extension */
export class AkatsukiExtension extends CDOTA_BaseNPC {

}

/** @extension */
class CapturedPreview extends CDOTA_BaseNPC {
    bidjuIndicatorParticle: ParticleID | null;
}

/** @extension */
export class BidjuDefender extends CDOTA_BaseNPC {
    bidjuName: string;
}


export class AkatsukiManager {
    static IsAkatsuki(unit: CDOTA_BaseNPC) {
        return unit.GetTeam() === DOTATeam_t.DOTA_TEAM_BADGUYS;
    }


    static OnBidjuCaptured(killedBidju: BidjuExtension) {
        AkatsukiManager.createParticleIndicator(killedBidju);
        AkatsukiManager.createDefender(killedBidju);
        if(killedBidju.owner){
            DebugPrint("capture this shit!")
        }
    }

    static createParticleIndicator(killedBidju: BidjuExtension) {
        const targetEntity = Entities.FindByName(undefined, `akatsuki_preview_${killedBidju.GetUnitName()}`) as CapturedPreview;
        if (targetEntity.bidjuIndicatorParticle) {
            ParticleManager.DestroyParticle(targetEntity.bidjuIndicatorParticle, false)
        }

        targetEntity.bidjuIndicatorParticle = ParticleManager.CreateParticle(
            "particles/env/bidju_capture_indicator.vpcf",
            ParticleAttachment_t.PATTACH_ABSORIGIN,
            targetEntity
        );

        ParticleManager.SetParticleControlEnt(
            targetEntity.bidjuIndicatorParticle,
            0,
            targetEntity,
            ParticleAttachment_t.PATTACH_ABSORIGIN,
            "attach_hitloc",
            targetEntity.GetAbsOrigin(),
            true
        )
    }

    static removeParticleIndicator(killedBidjuName: string){
        const targetEntity = Entities.FindByName(undefined, `akatsuki_preview_${killedBidjuName}`) as CapturedPreview;
        if (targetEntity.bidjuIndicatorParticle) {
            ParticleManager.DestroyParticle(targetEntity.bidjuIndicatorParticle, false)
        }
    }


    private static createDefender(killedBidju: BidjuExtension) {
        const entity = Entities.FindByName(undefined, `akatsuki_preview_${killedBidju.GetUnitName()}`);
        const unit = CreateUnitByName(
            "npc_dota_custom_cave",
            entity.GetAbsOrigin(),
            true,
            null,
            null,
            DOTATeam_t.DOTA_TEAM_BADGUYS
        ) as BidjuDefender;
        unit.bidjuName = killedBidju.GetUnitName();
        if(killedBidju.owner){
            killedBidju.owner.onLoseBidju();
        }
    }
}